File: frontend/src/api/session.ts
/**
 * API client for session management.
 */

export interface SessionResponse {
  session_id: string;
  bpm: number;
  bars_per_turn: number;
  turns_per_player: number;
  record_duration: number;
  /** Turn retry budget; the backend rejects POST /retry beyond this count. */
  max_turn_retries: number;
  base_track_url: string;
}

export interface TurnData {
  turn_number: number;
  player: 'user' | 'ai';
  transcription?: string;
  lyrics?: string;
  audio_url?: string;
}

export interface SessionStatus {
  step:
    | 'idle'
    | 'awaiting_user'
    | 'recording'
    | 'transcribing'
    | 'generating_lyrics'
    | 'generating_audio'
    | 'playing_response'
    | 'judging'
    | 'complete'
    | 'error';
  current_turn: number;
  turns_per_player: number;
  turn_history: TurnData[];
  transcription?: string;
  lyrics?: string;
  ai_audio_url?: string;
  error?: string;
  retry_count?: number;
  winner?: string;
  judge_reason?: string;
}

/**
 * Error carrying the HTTP status of a failed API response, so callers can tell
 * a permanent failure (404 - session gone) from a transient one (500, network).
 */
export class ApiError extends Error {
  readonly status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = 'ApiError';
    this.status = status;
  }
}

const API_BASE = '/api/session';

/**
 * Shown when the API cannot be reached at all, as opposed to answering with a
 * problem. Two cases produce it: a rejected fetch (nothing on the other end)
 * and a 502/503/504, which comes from the proxy in front of the API rather
 * than from the API itself. A free backend that has been suspended or has
 * spun down looks exactly like this, and "Service Unavailable" is not
 * something a player should be asked to read.
 */
const ARENA_DOWN_MESSAGE =
  'The arena is closed right now — the battle server is down. Try again later.';

function isArenaDown(status: number): boolean {
  return status === 502 || status === 503 || status === 504;
}

/** fetch(), but a transport failure reads as "the arena is closed". */
async function reachApi(input: string, init?: RequestInit): Promise<Response> {
  try {
    return await fetch(input, init);
  } catch {
    throw new Error(ARENA_DOWN_MESSAGE);
  }
}

/**
 * The AI opponent's character sheet, in the snake_case shape the API expects.
 * Every field is optional server-side, and the whole object may be omitted -
 * a session created without one just battles a nameless MC.
 */
export interface OpponentPersonaPayload {
  name?: string;
  age?: number;
  claims?: string;
  reality?: string;
  extra_info?: string;
}

export async function createSession(
  opponent?: OpponentPersonaPayload
): Promise<SessionResponse> {
  const response = await reachApi(API_BASE, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    // `opponent_name` is sent alongside the persona so older/simpler clients
    // (and the backend's fallback path) keep working unchanged.
    body: JSON.stringify({ opponent_name: opponent?.name, opponent }),
  });

  if (!response.ok) {
    if (response.status === 429) {
      const err = await response.json().catch(() => ({ detail: 'Rate limit exceeded' }));
      throw new Error(err.detail || 'Too many battles. Try again later.');
    }
    if (isArenaDown(response.status)) throw new Error(ARENA_DOWN_MESSAGE);
    throw new Error(`Failed to create session: ${response.statusText}`);
  }

  return response.json();
}

export async function uploadRecording(sessionId: string, audioBlob: Blob): Promise<void> {
  const formData = new FormData();
  formData.append('audio', audioBlob, 'recording.webm');

  const response = await reachApi(`${API_BASE}/${sessionId}/recording`, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    if (isArenaDown(response.status)) throw new Error(ARENA_DOWN_MESSAGE);
    const error = await response.json().catch(() => ({ detail: response.statusText }));
    throw new Error(error.detail || 'Failed to upload recording');
  }
}

export async function getSessionStatus(sessionId: string): Promise<SessionStatus> {
  const response = await fetch(`${API_BASE}/${sessionId}/status`);

  if (!response.ok) {
    const error = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new ApiError(
      error?.detail || `Failed to get session status: ${response.statusText}`,
      response.status
    );
  }

  return response.json();
}

export async function retryTurn(sessionId: string): Promise<{ status: string; retry_count?: number; message?: string }> {
  const response = await reachApi(`${API_BASE}/${sessionId}/retry`, {
    method: 'POST',
  });

  if (!response.ok) {
    if (isArenaDown(response.status)) throw new Error(ARENA_DOWN_MESSAGE);
    const error = await response.json().catch(() => ({ detail: response.statusText }));
    throw new Error(error.detail || 'Failed to retry');
  }

  return response.json();
}

File: api/models/session.py
"""Pydantic models for session management."""

from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field, field_validator, model_validator

from core.prompts.persona import (
    MAX_AGE,
    MAX_PERSONA_BLOCK_CHARS,
    MIN_AGE,
    PERSONA_FIELD_LIMITS,
    sanitize_prompt_text,
)


class PipelineStep(str, Enum):
    """Pipeline processing steps."""
    IDLE = "idle"
    AWAITING_USER = "awaiting_user"  # Waiting for user to record
    RECORDING = "recording"
    TRANSCRIBING = "transcribing"
    GENERATING_LYRICS = "generating_lyrics"
    GENERATING_AUDIO = "generating_audio"
    PLAYING_RESPONSE = "playing_response"
    JUDGING = "judging"
    COMPLETE = "complete"
    ERROR = "error"


class TurnData(BaseModel):
    """Data for a single turn in the battle."""
    turn_number: int
    player: Literal["user", "ai"]
    transcription: str | None = None  # User turns
    lyrics: str | None = None  # AI turns
    audio_url: str | None = None  # AI turns
    timing: dict[str, float] | None = None  # Per-stage seconds for AI turns


class OpponentPersona(BaseModel):
    """Client-supplied character sheet for the AI MC.

    UNTRUSTED. `POST /api/session` accepts any JSON body from anyone, so the
    frontend's OPPONENTS list is not a constraint on what arrives here. Each
    field caps the RAW value (over-length input is rejected with a 422) and is
    then sanitised down to a single harmless line before it can reach a prompt.
    Unknown keys (`id`, `imageSrc`, ...) are ignored rather than carried along.
    """
    name: str | None = Field(default=None, max_length=PERSONA_FIELD_LIMITS["name"])
    age: int | None = Field(default=None, ge=MIN_AGE, le=MAX_AGE)
    claims: str | None = Field(default=None, max_length=PERSONA_FIELD_LIMITS["claims"])
    reality: str | None = Field(default=None, max_length=PERSONA_FIELD_LIMITS["reality"])
    extra_info: str | None = Field(
        default=None, max_length=PERSONA_FIELD_LIMITS["extra_info"]
    )

    @field_validator("name", "claims", "reality", "extra_info")
    @classmethod
    def _sanitize(cls, value: str | None, info) -> str | None:
        if value is None:
            return None
        # Runs after the max_length constraint, so this only ever shortens a
        # value that was already within its cap.
        return sanitize_prompt_text(value, PERSONA_FIELD_LIMITS[info.field_name]) or None

    @model_validator(mode="after")
    def _cap_total_size(self) -> "OpponentPersona":
        total = sum(
            len(v)
            for v in (self.name, self.claims, self.reality, self.extra_info)
            if v
        )
        if total > MAX_PERSONA_BLOCK_CHARS:
            raise ValueError(
                f"opponent persona too long: {total} characters across all fields "
                f"(max {MAX_PERSONA_BLOCK_CHARS})"
            )
        return self


class SessionCreate(BaseModel):
    """Request model for creating a new session.

    Both fields are optional: a body of `{}` or a bare `opponent_name` is still
    a valid request, and the battle then runs without a persona.
    """
    opponent_name: str | None = Field(
        default=None, max_length=PERSONA_FIELD_LIMITS["name"]
    )
    opponent: OpponentPersona | None = None

    @field_validator("opponent_name")
    @classmethod
    def _sanitize_name(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return sanitize_prompt_text(value, PERSONA_FIELD_LIMITS["name"]) or None


class SessionResponse(BaseModel):
    """Response model for session creation."""
    session_id: str
    bpm: int
    bars_per_turn: int
    turns_per_player: int
    record_duration: int
    # The turn retry budget, so the frontend never has to hardcode a mirror
    # of the backend's MAX_TURN_RETRIES.
    max_turn_retries: int
    base_track_url: str


class SessionStatus(BaseModel):
    """Response model for session status polling."""
    step: PipelineStep
    current_turn: int
    turns_per_player: int
    turn_history: list[TurnData]
    transcription: str | None = None
    lyrics: str | None = None
    ai_audio_url: str | None = None
    error: str | None = None
    retry_count: int = 0
    timing: dict[str, float] | None = None  # Per-stage seconds for current turn
    winner: Literal["user", "ai", "draw"] | None = None
    judge_reason: str | None = None
